src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -566,32 +566,89 @@ private ValueTask<FlushResult> WriteDataAndTrailersAsync(Http2Stream stream, in
+    private void FinishWritingHeadersUnsynchronized(int streamId, int payloadLength, HPackHeaderWriter.HeaderWriteResult writeResult)
+    {
+        Debug.Assert(payloadLength <= _maxFrameSize, "The initial payload lengths is written to _headerEncodingBuffer with size of _maxFrameSize");
+        var bufferSize = _maxFrameSize;
+        if (writeResult == HPackHeaderWriter.HeaderWriteResult.Done)
+        {
+            // Fast path, only a single HEADER frame.
             _outgoingFrame.PayloadLength = payloadLength;
-
-            if (done)
+            _outgoingFrame.HeadersFlags |= Http2HeadersFrameFlags.END_HEADERS;
+            WriteHeaderUnsynchronized();
+            _outputWriter.Write(_headerEncodingBuffer.AsSpan(0, payloadLength));
+            return;
+        }
+        else if (writeResult == HPackHeaderWriter.HeaderWriteResult.MoreHeaders)
+        {
+            _outgoingFrame.PayloadLength = payloadLength;
+            WriteHeaderUnsynchronized();
+            _outputWriter.Write(_headerEncodingBuffer.AsSpan(0, payloadLength));
+        }
+        else
+        {
+            // This may happen in case of the TRAILERS after the initial encode operation.
+            while (writeResult == HPackHeaderWriter.HeaderWriteResult.BufferTooSmall)
             {
-                _outgoingFrame.ContinuationFlags = Http2ContinuationFrameFlags.END_HEADERS;
+                Debug.Assert(payloadLength == 0, "Payload written even though buffer is too small");
+                bufferSize *= 2;
+                var headerBuffer = ArrayPool<byte>.Shared.Rent(bufferSize);
+                var buffer = headerBuffer.AsSpan(0, bufferSize);
+                writeResult = HPackHeaderWriter.RetryBeginEncodeHeaders(_hpackEncoder, _headersEnumerator, buffer, out payloadLength);
+                if (writeResult != HPackHeaderWriter.HeaderWriteResult.BufferTooSmall)
+                {
+                    SplitHeaderFramesToOutput(streamId, buffer[..payloadLength], endOfHeaders: writeResult == HPackHeaderWriter.HeaderWriteResult.Done, isFramePrepared: true);
+                }
+                ArrayPool<byte>.Shared.Return(headerBuffer);
             }
+        }
 
-            WriteHeaderUnsynchronized();
-            _outputWriter.Write(buffer.Slice(0, payloadLength));
+        bufferSize = _maxFrameSize;
+
+        // HEADERS and zero or more CONTINUATIONS sent - all subsequent frames are (unprepared) CONTINUATIONs
+        while (writeResult != HPackHeaderWriter.HeaderWriteResult.Done)
+        {
+            byte[]? headerBuffer = null;
+            Span<byte> buffer = bufferSize == _maxFrameSize ? _headerEncodingBuffer : headerBuffer = ArrayPool<byte>.Shared.Rent(bufferSize);
+            writeResult = HPackHeaderWriter.ContinueEncodeHeaders(_hpackEncoder, _headersEnumerator, buffer[..bufferSize], out payloadLength);
+            if (writeResult == HPackHeaderWriter.HeaderWriteResult.BufferTooSmall)
+            {
+                bufferSize *= 2;
```

Comment:
Note to myself: maybe this should only rent an array here, instead of every iteration.